tests/integration/veth_test.py

```diff
@@ -0,0 +1,138 @@
+
+
+@pytest.mark.skipif(
+    nm_major_minor_version() >= 1.28,
+    reason="Modifying veth interfaces is supported on NetworkManager.",
+)
+def test_add_veth_not_supported():
+    desired_state = {
+        Interface.KEY: [
+            {
+                Interface.NAME: VETH1,
+                Interface.TYPE: InterfaceType.VETH,
+                Interface.STATE: InterfaceState.UP,
+                Veth.CONFIG_SUBTREE: {Veth.PEER: VETH1PEER},
+            }
+        ]
+    }
+
+    with pytest.raises(NmstateLibnmError):
+        libnmstate.apply(desired_state)
+
+
+@pytest.mark.skipif(
+    nm_major_minor_version() <= 1.28,
+    reason="Modifying veth interfaces is not supported on NetworkManager.",
+)
+def test_add_and_remove_veth():
+    with veth_interface(VETH1, VETH1PEER) as desired_state:
+        assertlib.assert_state(desired_state)
+
+    assertlib.assert_absent(VETH1)
+    assertlib.assert_absent(VETH1PEER)
+
+
+@pytest.mark.skipif(
+    nm_major_minor_version() <= 1.28,
+    reason="Modifying veth interfaces is not supported on NetworkManager.",
+)
+def test_add_veth_as_bridge_port():
+    with veth_interface(VETH1, VETH1PEER):
+        with bridges_with_port() as desired_state:
+            assertlib.assert_state_match(desired_state)
+
+
+@contextmanager
+def bridges_with_port():
+    d_state = {
+        Interface.KEY: [
+            {
+                Interface.NAME: "ovs-br0",
+                Interface.TYPE: InterfaceType.OVS_BRIDGE,
+                Interface.STATE: InterfaceState.UP,
+                Bridge.CONFIG_SUBTREE: {
+                    Bridge.PORT_SUBTREE: [
+                        {Bridge.Port.NAME: VETH1},
+                    ]
+                },
+            },
+            {
+                Interface.NAME: "br0",
+                Interface.TYPE: InterfaceType.LINUX_BRIDGE,
+                Interface.STATE: InterfaceState.UP,
+                Bridge.CONFIG_SUBTREE: {
+                    Bridge.PORT_SUBTREE: [
+                        {
+                            Bridge.Port.NAME: VETH1PEER,
+                        },
+                    ]
+                },
+            },
+        ]
+    }
+    try:
+        libnmstate.apply(d_state)
+        yield d_state
+    finally:
+        d_state[Interface.KEY][0][Interface.STATE] = InterfaceState.ABSENT
+        d_state[Interface.KEY][1][Interface.STATE] = InterfaceState.ABSENT
+        libnmstate.apply(d_state)
+
+
+@contextmanager
+def veth_interface(ifname, peer):
+    d_state = {
+        Interface.KEY: [
+            {
+                Interface.NAME: ifname,
+                Interface.TYPE: Veth.TYPE,
+                Interface.STATE: InterfaceState.UP,
+                Veth.CONFIG_SUBTREE: {
+                    Veth.PEER: peer,
+                },
+            }
+        ]
+    }
+    try:
+        libnmstate.apply(d_state)
+        yield d_state
+    finally:
+        d_state[Interface.KEY][0][Interface.STATE] = InterfaceState.ABSENT
```

Comment:
You should also remove the peer interface.

Reply:
When working with veth interfaces removing a veth interface it should remove the peer interface too. IMO, this is checking there is not a big bug here, if the tests fail because the peer still present means something very bad is happening.

Reply:
Done, thanks!